entry.go

```diff
@@ -126,43 +128,33 @@ func (te *TunaEntry) Start() {
 			var np *nkn.NanoPay
 			for {
 				time.Sleep(DefaultNanoPayUpdateInterval)
-				bytesIn := atomic.LoadUint64(&te.bytesIn)
-				bytesOut := atomic.LoadUint64(&te.bytesOut)
-				entryToExitPrice, exitToEntryPrice := te.GetPrice()
-				delta := exitToEntryPrice*common.Fixed64(bytesIn-te.bytesInPaid)/TrafficUnit + entryToExitPrice*common.Fixed64(bytesOut-te.bytesOutPaid)/TrafficUnit
-				if delta == 0 {
-					continue
-				}
-				paymentReceiver := te.GetPaymentReceiver()
-				if np == nil || np.Address() != paymentReceiver {
-					var err error
-					np, err = te.Wallet.NewNanoPay(paymentReceiver, te.config.NanoPayFee, DefaultNanoPayDuration)
-					if err != nil {
-						continue
-					}
-				}
-				tx, err := np.IncrementAmount(delta.String())
+				session, err := te.getSession(false)
 				if err != nil {
+					log.Printf("get session err: %v", err)
 					continue
 				}
-				txData := tx.ToArray()
-				session, err := te.getSession(false)
-				if err != nil {
+				bytesIn := atomic.LoadUint64(&te.bytesIn)
+				bytesOut := atomic.LoadUint64(&te.bytesOut)
+				entryToExitPrice, exitToEntryPrice := te.GetPrice()
+				cost := exitToEntryPrice*common.Fixed64(bytesIn-te.bytesInPaid)/TrafficUnit + entryToExitPrice*common.Fixed64(bytesOut-te.bytesOutPaid)/TrafficUnit
+				if cost == 0 {
 					continue
 				}
-				stream, err := session.OpenStream()
+				length, err := sendNanoPayment(np, session, te.Wallet, &cost, te.Common, te.config.NanoPayFee)
 				if err != nil {
-					continue
+					log.Printf("send nano payment err: %v", err)
+					return
 				}
-				n, err := stream.Write(txData)
-				if n == len(txData) && err == nil {
+				if length > 0 {
 					te.bytesInPaid = bytesIn
 					te.bytesOutPaid = bytesOut
 				}
-				stream.Close()
+				_, ok := <-te.closeChan
```

Comment:
You should use
```
select {
case <-te.closeChan:
    return
default:
}
```
otherwise it's going to block.
Also you should put this at the beginning of for loop so we can check tuna close even if `cost == 0`